File: dropwizard-extra-common/src/main/scala/com/datasift/dropwizard/zookeeper/config/ZooKeeperConfiguration.scala
package com.datasift.dropwizard.zookeeper.config

import com.yammer.dropwizard.config.Configuration
import org.hibernate.validator.constraints.{NotEmpty, Range}
import reflect.BeanProperty
import javax.validation.constraints.{Min, NotNull}
import com.yammer.dropwizard.util.Duration

/** Configuration for a ZooKeeper cluster */
class ZooKeeperConfiguration extends Configuration {

  /** hosts that make up the ZooKeeper quorum */
  @BeanProperty
  @NotEmpty
  val hosts = Array("localhost")

  /** port to connect to ZooKeeper nodes on */
  @BeanProperty
  @NotNull
  @Range(min = 0, max = 49151)
  val port = 2181

  /** timeout for idle connections to a ZooKeeper node */
  @BeanProperty
  @NotNull
  val timeout = Duration.seconds(60)

  /** Quorum specificiation as a comma-delimited String of host:port pairs. */
  def quorumSpec: String = hosts map (_ + ":" + port) mkString (",")
}

File: dropwizard-extra-core/src/main/java/com/datasift/dropwizard/config/ZooKeeperConfiguration.java
package com.datasift.dropwizard.config;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.base.Joiner;
import com.yammer.dropwizard.util.Duration;
import org.hibernate.validator.constraints.NotEmpty;
import org.hibernate.validator.constraints.Range;

import javax.validation.constraints.NotNull;

/**
 * A {@link com.yammer.dropwizard.config.Configuration} for a ZooKeeper cluster.
 */
public class ZooKeeperConfiguration {

    /**
     * Hostnames of every node in the ZooKeeper quorum.
     */
    @JsonProperty
    @NotEmpty
    protected String[] hosts = new String[]{ "localhost" };

    /**
     * Port to connect to every ZooKeeper node in the quorum on.
     */
    @JsonProperty
    @Range(min = 0, max = 49151)
    protected int port = 2181;

    /**
     * Maximum time to wait for a successful connection to a node in the quorum.
     */
    @JsonProperty
    @NotNull
    protected Duration connectionTimeout = Duration.seconds(6);

    /**
     * Maximum time to allow a ZooKeeper session to remain idle before ending it.
     */
    @JsonProperty
    @NotNull
    protected Duration sessionTimeout = Duration.seconds(6);

    /**
     * @see ZooKeeperConfiguration#hosts
     */
    public String[] getHosts() {
        return hosts;
    }

    /**
     * @see ZooKeeperConfiguration#port
     */
    public int getPort() {
        return port;
    }

    /**
     * @see ZooKeeperConfiguration#connectionTimeout
     */
    public Duration getConnectionTimeout() {
        return connectionTimeout;
    }

    /**
     * @see ZooKeeperConfiguration#sessionTimeout
     */
    public Duration getSessionTimeout() {
        return sessionTimeout;
    }

    /**
     * Retrieves a formatted specification of the ZooKeeper quroum.
     * <p>
     * The specification is formatted as: host1:port,host2:port[,hostN:port]
     *
     * @return a specification of the ZooKeeper quorum, formatted as a String
     */
    public String getQuorumSpec() {
        return Joiner
                .on(":" + getPort() + ",")
                .skipNulls()
                .appendTo(new StringBuilder(), getHosts())
                .append(':')
                .append(getPort())
                .toString();
    }
}
